### PR TITLE
EOS-19893: HA event analyzers IEM filter

### PR DESCRIPTION
**Pull request by shindeam**
## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    https://jts.seagate.com/browse/EOS-19893
    HA event analyzers IEM filter
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Need to implement HA event analyzers IEM filter
  </code>
</pre>
## Solution
<pre>
  <code>
    - Implement class IEMFilter which will filter the iem messages depend on the iem filter rules.
    - IEM filter rules will be our configuration stored in confstore
    - Filter class will have a method which will identify the msg is wither IEM or Alert.
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    ********IEM Filter********
    IEM filter test pass successfully
  </code>
</pre>

### CODE REVIEW

File: ha/core/event_analyzer/filter/filter.py [hunk 1]
```diff
@@ -86,3 +104,52 @@ def filter_event(self, msg: str) -> bool:
...
+
+class IEMFilter(Filter):
+    """ Filter IEM consumed by watcher """
+
+    def __init__(self):
+        """
+        Init method
+        """
+        super().__init__()
+        # Loads IEM filter rules in the configuration
```
Comment (ajay-paratmandali) [inline ha/core/event_analyzer/filter/filter.py:116]: Move this to `Filter` class as rule file same for both IEM and Alert

File: ha/core/event_analyzer/filter/filter.py [hunk 2]
```diff
@@ -41,17 +48,28 @@ def filter_event(self, msg: str) -> bool:
         """
         pass
 
+    @staticmethod
+    def is_iem_msg(msg: dict) -> bool:
```
Comment (ajay-paratmandali) [inline ha/core/event_analyzer/filter/filter.py:52]: is_iem_msg instade can we use it 
get_msg_type
to support more format and can be use by alert if needed

File: ha/core/event_analyzer/filter/filter.py [hunk 3]
```diff
@@ -86,3 +104,52 @@ def filter_event(self, msg: str) -> bool:
...
+            if msg_type is not None:
+                return iem_required
+
+            if not Filter.is_iem_msg(message):
+                return iem_required
+
+            msg_type = message['message']['sensor_response_type']
+            _module_type = msg_type['specific_info']['module']
+
+            if self.filter_type == const.INCLUSION:
```
Comment (ajay-paratmandali) [inline ha/core/event_analyzer/filter/filter.py:143]: we also need to check component should be in list

also can we modify rule as 
```
    "iem": {
        "filter_type":"inclusion",
        "components": ["ha"],
        "ha": {
            "module": ["Node", "Service"]
        }
    }
```
first match component then module

**Comment by indrajitzagade** [comment]
@shindeam Please fix DCO check.

File: ha/core/event_analyzer/filter/filter.py (hunk 2 above)
Comment (indrajitzagade) [reply ha/core/event_analyzer/filter/filter.py:52]: Agree

File: ha/core/event_analyzer/filter/filter.py (hunk 1 above)
Comment (indrajitzagade) [reply ha/core/event_analyzer/filter/filter.py:116]: Agree

Comment (shindeam) [reply ha/core/event_analyzer/filter/filter.py]: Yes, done.

File: ha/core/event_analyzer/filter/filter.py (hunk 2 above)
Comment (shindeam) [reply ha/core/event_analyzer/filter/filter.py]: Done.

File: ha/core/event_analyzer/filter/filter.py (hunk 3 above)
Comment (shindeam) [reply ha/core/event_analyzer/filter/filter.py:142]: Done, updated the iem filter rules format
"iem": {
    "filter_type":"inclusion",
    "components": ["ha"],
    "modules": {"ha": ["Node", "Service"] }
}

**Comment by shindeam** [comment]
> @shindeam Please fix DCO check.

Yes updated.

File: ha/core/event_analyzer/filter/filter.py [hunk 4]
```diff
@@ -86,3 +102,54 @@ def filter_event(self, msg: str) -> bool:
...
+
+            _msg_type = Filter.get_msg_type(message)
+            if _msg_type != MESSAGETYPE.IEM.value:
+                return iem_required
+
+            _sensor_response_type = message.get('sensor_response_type')
+            _component_type = _sensor_response_type['specific_info']['module']
+            _module_type = _sensor_response_type['specific_info']['module']
+
+            if self.filter_type == const.INCLUSION:
```
Comment (ajay-paratmandali) [inline ha/core/event_analyzer/filter/filter.py:142]: if self.filter_type == const.INCLUSION:

if self.filter_type == const.EXCLUSION:

both must be if else

**Review by ajay-paratmandali — APPROVED** [review]

File: ha/core/event_analyzer/filter/filter.py [hunk 5]
```diff
@@ -86,3 +102,54 @@ def filter_event(self, msg: str) -> bool:
...
+                if _component_type in self.components_list and _module_type in self.modules_dict.get(_component_type):
+                    iem_required = True
+            elif self.filter_type == const.EXCLUSION:
+                if _component_type not in self.components_list or _module_type not in self.modules_dict.get(
+                        _component_type):
+                    iem_required = True
+            else:
```
Comment (ajay-paratmandali) [inline ha/core/event_analyzer/filter/filter.py:149]: `Log.error("Invalid IEM filter type in the event IEM filter rules")`

This case should not exist 
it is validation check
can you check it in validate() function and call validate_rule in init of this class

Comment (shindeam) [reply ha/core/event_analyzer/filter/filter.py]: Done.

File: ha/const.py [hunk 6]
```diff
@@ -151,6 +151,8 @@
 SOURCE_ALERT_FILTER_RULES_FILE = "{}/alert_filter_rules.json".format(SOURCE_CONFIG_PATH)
 SYSTEM_SERVICE_FILE = "{}/event_analyzer.service".format(SYSTEM_DIR)
 SOURCE_SERVICE_FILE = "{}/conf/service/event_analyzer.service".format(SOURCE_PATH)
+IEM_DESCRIPTION="WS0080010001, Node, The cluster has lost $host server. System is running in degraded mode. " \
```
Comment (indrajitzagade) [inline ha/const.py:154]: I guess this is a temporary addition and will be removed after the alert generator task adds this to a config file.

Comment (shindeam) [reply ha/const.py:154]: Yes, this is a temporary addition for now.

**Review by indrajitzagade — APPROVED** [review]

File: ha/core/event_analyzer/filter/filter.py [hunk 7]
```diff
@@ -41,17 +49,25 @@ def filter_event(self, msg: str) -> bool:
         """
         pass
 
+    @staticmethod
+    def get_msg_type(msg: dict) -> str:
+        """
+        Check if the msg type is iem or alert
+        """
+        msg_type = msg.get("sensor_response_type")
+        return msg_type["info"]["resource_type"]
```
Comment (ajaysrivas) [inline ha/core/event_analyzer/filter/filter.py:58]: Move these literals to const. Check other places too.
Create a class for AlertConstants and IemConstants.

Comment (shindeam) [reply ha/core/event_analyzer/filter/filter.py]: Done.

File: ha/const.py [hunk 8]
```diff
@@ -151,6 +151,15 @@
 SOURCE_ALERT_FILTER_RULES_FILE = "{}/alert_filter_rules.json".format(SOURCE_CONFIG_PATH)
 SYSTEM_SERVICE_FILE = "{}/event_analyzer.service".format(SYSTEM_DIR)
 SOURCE_SERVICE_FILE = "{}/conf/service/event_analyzer.service".format(SOURCE_PATH)
+ACTUATOR_RESPONSE_TYPE= "actuator_response_type"
```
Comment (ajaysrivas) [inline ha/const.py:154]: These can also go to AlertConstants Class as these are related to alert?

Comment (shindeam) [reply ha/const.py:154]: Actually those are common so kept in const file only instead of alert or iem constants.

**Comment by ajaysrivas** [comment]
Please do take care of arranging constants as done in https://github.com/Seagate/cortx-ha/pull/408. Merging it for now.

**Pull request merged**